Homework 1

BulatZuraev opened the pull request.

BulatZuraev committed:
implement stmt interpreter

BulatZuraev committed:
implement stmt to sm compiler
implement sm interpreter

BulatZuraev committed:
fixes

BulatZuraev committed:
fixes

dboulytchev commented:
Такой код зачесть не могу:

1. Зачем везде "return"?
2. Во всех таких местах

                    local valWorld, st;
                    valWorld := c.snd.readWorld;
                    st := c.fst <- [var, valWorld.fst];
                    return [st, valWorld.snd]

     а. зачем дополнительные локальные переменные?
     b. зачем присваивания?
     c. вместо  snd/fst нужно использовать сопоставление с образцом.

3. Интерпретатор стековой машины должен получать список инструкций, а у Вас он получает список списков (и делает затем deepFlatten). Аналогично, компилятор для стековой машины должен возвращать список инструкций, а он возвращает список списков.

BulatZuraev committed:
remove returns
change compileSM and compileExpr return value
replace fst and snd calls with pattern matching

BulatZuraev commented:
Исправил

danyaberezun commented:
Зачтено, хотя 'deepFlatten' всё ещё квадратичным делает решение.
